Add initial barcode validation and tests

harikt opened the pull request:
Hi @weierophinney ,

This PR contains Barcode validation .

Some of the tests are commented out which need to be addressed. I am still thinking how we can resolve those.

- [x] Are you creating a new feature?
  - [ ] Why is the new feature needed? What purpose does it serve?
  - [x] Add tests for the new feature.
  - [x] Add documentation for the new feature.
  - [ ] Add a `CHANGELOG.md` entry for the new feature.

harikt commented on src/Barcode.php:
```diff
@@ -0,0 +1,159 @@
+<?php
+declare(strict_types=1);
```
I wonder why `declare(strict_types=1);` is omitted on other classes. Is that purposeful ?

weierophinney replied:
We haven't yet decided if we will require that.

harikt commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,530 @@
+    // {
+    //     $this->expectException(InvalidArgumentException::class);
+    //     $this->expectExceptionMessage('not found');
+    //     $barcode = new Barcode('\Zend\Validate\BarcodeTest\NonExistentClassName');
```
In the old code this send an InvalidArgumentException before calling `getAdapter()`  . Do we want to do the same ? In that case the adapter is created first even if it may / may not is used.

harikt commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,531 @@
+
+
+
+    // public function testDefaultArrayConstructWithMissingAdapter()
```
I think we can remove these tests. What do you think ?

weierophinney replied:
Yes, as these ports should already the constructor to accept specific arguments, not an option array.

harikt commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,531 @@
+        $barcode = new Barcode('ean8');
+        $result = $barcode->validate('123');
+        $this->validateResult($result, false);
+        // $message = $barcode->getMessages();
```
`$result->getMessages()` returns array of `ValidationFailureMessage` . If we can keep the key as `code` of `ValidationFailureMessage` it may be easy to check this. May be there is another way to resolve this.

weierophinney replied:
The code is part of the validation failure messages.

weierophinney commented:
@,hatikt - Thanks for starting this! However, per the RFC, this is one of the validators that will get a bridge component, due to the required dependency it had on zend-barcode. If you want to create that component, you can do so under your own user/org, and then open an issue here (or, once we've moved the component to the zendframework org, on that repo) letting us know, so we can bring it into the zendframework org.

The new component would depend on this one and zend-barcode.

harikt commented:
Hi @weierophinney , 

Thank you for your quick reply. I am a bit confused for I didn't noticed any dependency on zend-barcode when writing a test for the Barcode validation component.

I will look more into the discussion on https://github.com/zendframework/zend-validator/issues/1 and https://discourse.zendframework.com/t/rfc-new-validation-component/208/2 .

weierophinney commented:
> I am a bit confused for I didn't noticed any dependency on zend-barcode when writing a test for the Barcode validation component.

Okay, so it turns out I was wrong! In reviewing the various barcode-related validators, none of them have a dependency on zend-barcode. Additionally, it's not a suggested (or required!) dependency either.

So... keep up the great work, and I'll review when you've finished!

harikt commented:
@weierophinney I have updated the code. I belive you can do a final review.

weierophinney commented on src/Barcode/Royalmail.php:
```diff
@@ -0,0 +1,93 @@
+            } else {
+                return false;
+            }
+        }
```
Let's rewrite the above:

```php
if ($value[0] === '(') {
    if ($value[strlen($value) - 1] !== ')') {
        return false;
    }

    $value = substr($value, 1, -1);
}
```

weierophinney commented on src/Barcode/Upce.php:
```diff
@@ -0,0 +1,40 @@
+            $this->useChecksum(false);
+        } else {
+            $this->useChecksum(true);
+        }
```
This can be much simpler:

```php
$this->setUseChecksum(strlen($value) === 8);
```

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+use Zend\DataValidator\Exception\InvalidArgumentException;
+use Zend\DataValidator\Barcode\AdapterInterface;
+use Zend\DataValidator\Barcode\Ean13;
+use Zend\DataValidator\Barcode\Issn;
```
Please alphabetize these.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+ * \Zend\Barcode
+ *
+ * @group      Zend_Validator
+ */
```
This docblock is irrelevant, and can be removed.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+
+    private function validateResult(Result $result, $expectedResult)
+    {
+        $this->assertInstanceOf(Result::class, $result);
```
This assertion is unnecessary, due to the typehint; PHP will raise an error for an invalid `$result`, which will cause a failed test.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+        $this->assertSame(
+            $expectedResult,
+            $result->isValid(),
+            'Failed value: ' . var_export($result->getValue(), true)
```
Let's make this more descriptive:

```php
sprintf(
    'Expected validation result "%s" was not received; value validated: %s',
    var_export($expectedResult, true),
    var_export($result->getValue(), true)
)
```

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+    public function provideBarcodeConstructor()
+    {
+        return [
+            'null' => [null, Ean13::class],
```
This test case will need to be removed once the `AdapterInterface` typehint is added to the `$adapter` constructor argument. As such, the entire data provider can likely be removed, and, instead, the `testBarcodeConstructor()` test case can use a mock instance of `AdapterInterface` to test it.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+        $this->expectExceptionMessage('not found');
+        $barcode = new Barcode('\Zend\Validate\BarcodeTest\NonExistentClassName');
+        $barcode->getAdapter();
+    }
```
This test case can be removed once the updates to the constructor are made.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+        $barcode = new Barcode($this->createMock(AdapterInterface::class));
+
+        $this->assertInstanceOf(AdapterInterface::class, $barcode->getAdapter());
+    }
```
This test will duplicate the `testBarcodeConstructor()` test once the constructor argument is changed, and can thus be removed.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+            'non-string-validation-upca-array'  => ['upca', ['065100004327'], false],
+            'non-string-validation-ean13-int'   => ['ean13', 106510000.4327, false],
+            'non-string-validation-ean13-array' => ['ean13', ['065100004327'], false],
+        ];
```
This provider should be updated to provide actual instances of the adapters.

weierophinney commented on test/_files/MyBarcode1.php:
```diff
@@ -0,0 +1,20 @@
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator\Barcode;
```
Use the `ZendTest\DataValidator\Barcode` namespace, and then import the `AbstractAdapter`.

weierophinney commented on test/_files/MyBarcode2.php:
```diff
@@ -0,0 +1,20 @@
+
+namespace Zend\DataValidator\Barcode;
+
+class MyBarcode2 extends AbstractAdapter
```
Same comment applies here.

weierophinney commented on test/_files/MyBarcode3.php:
```diff
@@ -0,0 +1,20 @@
+ * @license   https://github.com/zendframework/zend-datavalidator/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\DataValidator\Barcode;
```
and here, and with all of these test adapters.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+
+    public function testInvalidChecksumAdapter()
+    {
+        require_once __DIR__ . "/_files/MyBarcode1.php";
```
These `require_once` calls can go away once you update the various test adapters to use the `ZendTest` namespace. Doing so will require adding a new `autoload-dev` rule to `composer.json` so they may be autoloaded.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+     * @dataProvider validationProvider
+     */
+    public function testValidateReturnsExpectedResults(
+        $adapter,
```
Typehint this argument.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+    public function testInvalidChecksumAdapter()
+    {
+        require_once __DIR__ . "/_files/MyBarcode1.php";
+        $barcode = new Barcode('MyBarcode1');
```
Here and throughout this file, you'll need to switch from using string "short" names to `new <CLASSNAME>`.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+        $this->expectExceptionMessage('does not implement');
+        require_once __DIR__ . "/_files/MyBarcode5.php";
+        $barcode->setAdapter('MyBarcode5');
+    }
```
This method can be removed after the constructor changes are made, and the `setAdapter()` method is removed.

weierophinney commented on test/BarcodeTest.php:
```diff
@@ -0,0 +1,514 @@
+        $this->validateResult($barcode->validate('123'), true);
+        $this->validateResult($barcode->validate('123a'), false);
+
+        $barcode->useChecksum(true);
```
Once the change to `useChecksum()` is made in the `Barcode` class, you'll instead change the flag on the _adapter_ within these tests.

harikt commented:
Thank you for the detailed review @weierophinney . I will send an update in the coming days.

harikt commented on src/Barcode.php:
```diff
@@ -0,0 +1,163 @@
+
+            if (! is_null($this->useChecksum)) {
+                $adapter->useChecksum($this->useChecksum);
+            }
```
Hi @weierophinney , 

If we are passing the concrete adapter, do we really need to pass the `$useChecksum` . There is no place where this variable need to be used. Can't we remove this ?

weierophinney replied:
If the main barcode validator doesn't and/or shouldn't override what's in the adapter, then we can remove it.

harikt commented:
Hi @weierophinney ,

Have updated code as requested and added documentation ( copied from old barcode validator , updated with the current ones )

Thank you.

I will add the documentation

harikt commented on src/Bitwise.php:
```diff
@@ -0,0 +1,183 @@
+
+use Traversable;
+
+class Bitwise extends AbstractValidator
```
Oh shit, I committed this file accidentally 👎  . @weierophinney what do you want to do about this ?

Remove this file from this PR and send a different PR . 

or add unit test / fix for this Bitwise one ?

weierophinney replied:
Remove it.

I've taken the liberty of rebasing your branch off of current master, and I did an interactive rebase to remove the `Bitwise` class file from the commit where it was introduced.

harikt replied:
So do I want to do anything here ? I am a bit confused with the sentence.

> I've taken the liberty of rebasing your branch off of current master, and I did an interactive rebase to remove the Bitwise class file from the commit where it was introduced.

So did you removed the `Bitwise` class or do you want me to remove it ? or when you are doing the rebasing you will do ?

harikt commented:
It looks like you have removed the Bitwise class. Is there anything else that makes it a blocker for not merging ? I got only one email, but this one still shows requested changes. Not sure what are the changes requested newly. I assume all changes have been already integrated earlier itself. Let me know if I have missed anything.

weierophinney commented:
@harikt Sorry for the delay getting back to you!

I'd started on some refactoring, but then had other obligations last week that prevented me from finishing, as well as providing you details of what I was doing.

What I've done, in general, is as follows:

- Removed all public setters, in favor of constructor arguments. This ensures that the various barcode adapters, and the `Barcode` validator itself, remain stateless (one of the primary goals of the entire component!).
- I've separated out the "checksum" functionality into an additional interface and trait. Adapters that provide checksum capabilities now implement that interface (optionally via the trait). The `Barcode` validator checks to see if the adapter implements that interface, and if a checksum validation is required, before attempting to validate the checksum. Checksum methods were moved either into the specific adapters, or, if used in more than one location, into traits. This change greatly simplified the `AbstractAdapter`.
- Added `declare(strict_types=1)` throughout all new classes. This helped me identify and correct some problems with length, character, and checksum validations that have been lingering since they were originally written.

I've also updated the documentation to list out any constructor arguments, and to push the usage example to the top of the page.

Thanks a ton for your work on this!

harikt commented:
> Sorry for the delay getting back to you!

No issues. 

The problem with too much delay is we all will be carried away to other stuffs and forget about what we did and why we did ( Probably a failure at my end to document things ) . I always forget in 2-3 days and find it hard to get back to it. This is not a complaint. Just telling about my personal thing.

And thank you for your good work here. Any plans about release date etc ?
